### PR TITLE
feat: jwt-auth support extension payload

### LINKED ISSUE #3594: how to set the JWT content as header values?
### Issue description
Hi,
I need apisix to validate a JWT token in the request header that was issued by my own auth-service and set the content of the JWT token (like user_id) as header arguments. I've read through the JWT doc section, but I can't find how to set the JWT content as header values.
### Environment

* apisix version (cmd: `apisix version`): latest
* OS (cmd: `uname -a`): ubuntu
* OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`): latest
* etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
* apisix-dashboard version, if have:

### PR DESCRIPTION
**Pull request by benx203**
### What this PR does / why we need it:

Make jwt-auth support extension payload.
Fix #3594 

### Pre-submission checklist:

* [X] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [X] Have you modified the corresponding document?
* [X] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/jwt-auth.lua [hunk 1]
```diff
@@ -104,7 +105,7 @@ do
         core.table.clear(consumer_names)
 
         for _, consumer in ipairs(consumers.nodes) do
-            core.log.info("consumer node: ", core.json.delay_encode(consumer))
+            log.info("consumer node: ", core.json.delay_encode(consumer))
```
Comment (spacewander) [inline apisix/plugins/jwt-auth.lua:108]: The original `core.log` works well. Please don't touch unrelative code.

File: apisix/plugins/jwt-auth.lua [hunk 2]
```diff
@@ -186,48 +187,55 @@ local function get_secret(conf)
...
 
 
-local function sign_jwt_with_HS(key, auth_conf)
+local function get_real_payload(key, auth_conf, payload)
+    local real_payload = {
+        key = key,
+        exp = ngx_time() + auth_conf.exp
+    }
+    if payload then
+        local payloadEx = core.json.decode(payload)
```
Comment (spacewander) [inline apisix/plugins/jwt-auth.lua:196]: Better to use `extra_payload`

File: apisix/plugins/jwt-auth.lua [hunk 3]
```diff
@@ -277,18 +285,44 @@ function _M.rewrite(conf, ctx)
...
         return 401, {message = jwt_obj.reason}
     end
 
     consumer_mod.attach_consumer(ctx, consumer, consumer_conf)
-    core.log.info("hit jwt-auth rewrite")
+    log.info("hit jwt-auth rewrite")
+end
+
+
+local function user_info()
```
Comment (spacewander) [inline apisix/plugins/jwt-auth.lua:303]: Is this function necessary for the core feature? We should only accept the necessary API to avoid bloating.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/plugins/jwt-auth.lua (hunk 3 above)
Comment (benx203) [reply apisix/plugins/jwt-auth.lua:303]: Yes,the function provide ability to check up the payload content.

File: apisix/plugins/jwt-auth.lua (hunk 1 above)
Comment (benx203) [reply apisix/plugins/jwt-auth.lua:108]: OK.

File: apisix/plugins/jwt-auth.lua (hunk 3 above)
Comment (spacewander) [reply apisix/plugins/jwt-auth.lua:302]: This feature can be implemented as a local script to check.
There is no need to expose it as a public API of the gateway.

Comment (benx203) [reply apisix/plugins/jwt-auth.lua]: I get it,thanks.

**Comment by spacewander** [comment]
Would you add two test cases with payload, one for default algorithm and another for RS256?

Like https://github.com/apache/apisix/blob/86e49837d20147471eeb852d60d4c5e2f15a2bd4/t/plugin/jwt-auth.t#L1180

Thanks!

File: t/plugin/jwt-auth.t [hunk 4]
```diff
@@ -1209,7 +1241,39 @@ hello world
 
 
 
-=== TEST 43: test for unsupported algorithm
+=== TEST 44: sign / verify (algorithm = HS512,with extra payload)
```
Comment (spacewander) [inline t/plugin/jwt-auth.t:1244]: Maybe you are misled by my example. We need two tests, one for default and another for RS256 (but not for HS512).

Comment (benx203) [reply t/plugin/jwt-auth.t:1244]: Test 19 is for RS256.

**Review by spacewander — APPROVED** [review]
Next time, please add new tests **at the bottom**.

**Comment by benx203** [comment]
> Next time, please add new tests **at the bottom**.

@spacewander OK,get it.

**Review by tokers — APPROVED** [review]

**Pull request merged**